packages/gatsby/src/schema/infer/add-inferred-fields.js

```diff
@@ -0,0 +1,321 @@
+
+module.exports = {
+  addInferredFields,
+}
+
+const addInferredFieldsImpl = ({
+  schemaComposer,
+  typeComposer,
+  nodeStore,
+  exampleObject,
+  typeMapping,
+  prefix,
+  depth,
+  addDefaultResolvers,
+}) => {
+  const fields = {}
+  Object.keys(exampleObject).forEach(unsanitizedKey => {
+    const exampleValue = exampleObject[unsanitizedKey]
+    let key = createFieldName(unsanitizedKey)
+    const selector = `${prefix}.${key}`
+
+    let arrays = 0
+    let value = exampleValue
+    while (Array.isArray(value)) {
+      value = value[0]
+      arrays++
+    }
+
+    let fieldConfig
+    if (hasMapping(typeMapping, selector)) {
+      // TODO: Use `prefix` instead of `selector` in hasMapping and getFromMapping?
+      // i.e. does the config contain sanitized field names?
+      fieldConfig = getFieldConfigFromMapping(typeMapping, selector)
+    } else if (key.includes(`___NODE`)) {
+      fieldConfig = getFieldConfigFromFieldNameConvention(
+        schemaComposer,
+        nodeStore,
+        exampleValue,
+        unsanitizedKey
+      )
+      key = key.split(`___NODE`)[0]
+    } else {
+      fieldConfig = getFieldConfig(
+        schemaComposer,
+        nodeStore,
+        value,
+        selector,
+        depth,
+        typeMapping
+      )
+    }
+
+    while (arrays--) {
+      fieldConfig = { ...fieldConfig, type: [fieldConfig.type] }
+    }
+
+    // Proxy resolver to unsanitized fieldName in case it contained invalid characters
+    if (key !== unsanitizedKey) {
+      // Don't create a field with the sanitized key if a field with that name already exists.
+      invariant(
+        exampleObject[key] == null && !typeComposer.hasField(key),
+        `Invalid key ${unsanitizedKey} on ${prefix}. GraphQL field names must ` +
+          `only contain characters matching /^[a-zA-Z][_a-zA-Z0-9]*$/. and ` +
+          `must not start with a double underscore.`
+      )
+
+      const resolver = fieldConfig.resolve || defaultFieldResolver
+      fieldConfig = {
+        ...fieldConfig,
+        resolve: (source, args, context, info) =>
+          resolver(source, args, context, {
+            ...info,
+            fieldName: unsanitizedKey,
+          }),
+      }
+    }
+
+    if (typeComposer.hasField(key)) {
+      let fieldType = typeComposer.getFieldType(key)
+      if (_.isPlainObject(value) /* && depth < MAX_DEPTH */) {
+        // TODO: Use helper (similar to dropTypeModifiers)
+        let lists = 0
+        while (fieldType.ofType) {
+          if (fieldType instanceof GraphQLList) lists++
+          fieldType = fieldType.ofType
+        }
+
+        if (lists === arrays && fieldType instanceof GraphQLObjectType) {
+          addInferredFieldsImpl({
+            schemaComposer,
+            typeComposer: typeComposer.getFieldTC(key),
+            exampleObject: value,
+            typeMapping,
+            nodeStore,
+            prefix: selector,
+            depth: depth + 1,
+            addDefaultResolvers: true,
```

Comment:
Why is addDefaultResolvers true here?

Reply:
Fixed to inherit it from parent.

Reply:
We probably also want `addDefaultResolvers` in `getFieldConfig`.